File: todo_list/py/todo_list.py
from datetime import datetime

PRIORITIES = {
    "IMPORTANT": 1,
    "DAILY": 2,
    "MAYBE": 3
}


class Task:
    _task_name = ""
    _due_date = None
    _priority = None

    @property
    def task_name(self):
        return self._task_name

    @task_name.setter
    def task_name(self, task_name):
        self._task_name = task_name

    @property
    def due_date(self):
        return self._due_date

    @due_date.setter
    def due_date(self, due_date):
        self._due_date = self._parse_due_date(due_date)

    @property
    def priority(self):
        return self._priority

    @priority.setter
    def priority(self, priority):
        if (priority < 1 or priority > 3):
            raise Exception("invalid priority range")

        self._priority = priority

    def __init__(self, task_name, due_date, priority):
        if (not task_name or not due_date or priority != 0 and not priority):
            raise Exception("missing obrigatory data")

        if (priority < 1 or priority > 3):
            raise Exception("invalid priority range")

        self._task_name = task_name
        self._priority = priority
        self._due_date = self._parse_due_date(due_date)

    def _parse_due_date(self, due_date):
        dateValues = due_date.split("/")

        if (len(dateValues) != 3):
            raise Exception("invalid due date")

        return datetime(
            int(dateValues[0]),
            int(dateValues[1]),
            int(dateValues[2])
        )

    def update_task_due_date_with_date_obj(self, newDue):
        if (not isinstance(newDue, datetime)):
            raise Exception("invalid object")

        self._due_date = newDue


class TodoList:
    _list = []

    @property
    def list(self):
        return [i for i in self._list]

    def __init__(self):
        self._list = []

    def add_task(self, task):
        if (not isinstance(task, Task)):
            raise Exception("invalid object")

        self._list.append(task)

    def remove_task(self, task_name):
        removed_tasks = []
        updated_tasks = []
        for task in self._list:
            if (task.task_name == task_name):
                removed_tasks.append(task)
                continue
            updated_tasks.append(task)

        self._list = updated_tasks
        return removed_tasks

    def update_task(self, task_name, updatedTask):
        if (not isinstance(updatedTask, Task)):
            raise Exception("invalid object")

        updated_count = 0

        for task in self._list:
            if (task.task_name == task_name):
                updated_count += 1
                task.task_name = updatedTask.task_name
                task.priority = updatedTask.priority
                task.update_task_due_date_with_date_obj(updatedTask.due_date)

        if (not updated_count):
            raise Exception("task not found")

        return updated_count

File: todo_list/js/todo_list.mjs
export const PRIORITIES = {
    IMPORTANT: 1,
    DAILY: 2,
    MAYBE: 3
}

export class Task {
    /** @type {string} */
    #taskName
    /** @type {Date} */
    #dueDate
    /** @type {number} */
    #priority

    /**
     * @param {string} taskName
     * @param {string} dueDate
     * @param {number} priority
     */
    constructor(taskName, dueDate, priority) {
        if (priority < 1 || priority > 3) {
            throw new Error("invalid priority range")
        }

        if (!taskName || !dueDate || !priority) {
            throw new Error("missing obrigatory data")
        }

        this.#taskName = taskName;
        this.#dueDate = this.#parseDueDate(dueDate);
        this.#priority = priority;
    }

    get taskName() {
        return this.#taskName
    }

    set taskName(taskName) {
        this.#taskName = String(taskName)
    }

    get dueDate() {
        return this.#dueDate
    }

    set dueDate(dueDate) {
        this.#dueDate = this.#parseDueDate(dueDate);
    }

    get priority() {
        return this.#priority
    }

    set priority(priority) {
        if (priority < 1 || priority > 3) {
            throw new Error("invalid priority range")
        }

        this.#priority = priority
    }

    /**
     * @param {string} dueDate - expects string in the format: yyyy/mm/dd
     * @returns {Date}
     */
    #parseDueDate(dueDate) {
        const [year, month, dd] = dueDate.split("/")

        if (!year || !month || !dd) {
            throw new Error("invalid due date")
        }

        return new Date(year, Number(month) - 1, dd)
    }

    /**
     * @param {Date} newDate
     */
     updateDueDateWithDateObj(newDate) {
         if (!(newDate instanceof Date)) {
             throw new Error("invalid object")
         }

         this.#dueDate = newDate
     }
}

export class TodoList {
    /** @type {Array<Task>} */
    #list = [];

    get list() {
        return [...this.#list]
    }

    /**
     * @param {Task} task
     */
    addTask(task) {
        if (!(task instanceof Task)) {
            throw new Error("invalid object")
        }

        this.#list.push(task)
    }

    /**
     * @param {string} taskName
     * @returns {Array<Task>}
     */
    removeTask(taskName) {
        const removedTasks = []
        const newList = []
        for (const task of this.#list) {
            if (task.taskName === taskName) {
                removedTasks.push(task)
                continue
            }
            newList.push(task)
        }
        this.#list = newList

        return removedTasks
    }

    /**
     * @param {string} taskName
     * @param {Task} updatedTask
     * @returns {number}
     */
    updateTask(taskName, updatedTask) {
        if (!(updatedTask instanceof Task)) {
            throw new Error("invalid object")
        }

        let updatedTasks = 0;

        for (const task of this.#list) {
            if (task.taskName === taskName) {
                updatedTasks++
                task.taskName = updatedTask.taskName
                task.priority = updatedTask.priority
                task.updateDueDateWithDateObj(updatedTask.dueDate)
            }
        }

        if (updatedTasks == 0) {
            throw new Error("task not found")
        }

        return updatedTasks
    }
}
